aws-cpp-sdk-core/source/utils/threading/Executor.cpp

```diff
@@ -107,3 +107,40 @@ bool PooledThreadExecutor::HasTasks()
     std::lock_guard<std::mutex> locker(m_queueLock);
     return m_tasks.size() > 0;
 }
+
+BlockingExecutor::BlockingExecutor(std::shared_ptr<Executor> executor, size_t poolSize) :
+m_executor(executor), m_poolSize(poolSize), m_numTasksRunning(0)
+{
+}
+
+BlockingExecutor::~BlockingExecutor()
+{
+}
+
+bool BlockingExecutor::SubmitToThread(std::function<void()>&& fn)
+{
+    std::unique_lock<std::recursive_mutex> locker(m_syncPointLock);
+    if (m_numTasksRunning >= m_poolSize)
+    {
+        m_syncPoint.wait(locker);
+    }
+    m_numTasksRunning++;
+    locker.unlock();
+    
+    m_executor->Submit(&BlockingExecutor::ExecuteTask, this, fn);
+    return true;
+}
+
+void BlockingExecutor::ExecuteTask(std::function<void()> fn)
+{
+    fn();
+    OnTaskComplete();
+}
+
+void BlockingExecutor::OnTaskComplete()
+{
+    m_syncPoint.notify_one();
+    std::unique_lock<std::recursive_mutex> locker(m_syncPointLock);

```

Comment:
Upon thought: you need to hold the locker around the notify_one() call to be 100% correct wrt ordering of tasks.

Moving this around we fixed a race where a new task could be submitted when a whole bunch of other tasks were still waiting (and this would cause m_numTasksRunning to be > m_poolSIze).

As coded the situation is much more subtle: m_numTasksRunning can easily be > m_poolSize but it does not reflect the ACTUAL number of guys running. And no new guys will be allowed in.

Might be easiest to just move 142 below the locker and be done with it -- and then you can also assert that m_numTasksRunning is <= m_poolSize.

It's also good to write a test (a stress test) which just adds 100s of tasks with m_poolSize=10 just to make sure they all complete and there is no crash.